Big-O notasyonu, algoritmanın çalışmasının nasıl ölçeklendiğini anlatır ve temel olarak şu sorulara cevap arar:

- **Zaman Karmaşıklığı (Time Complexity)**: Algoritmanın çalışma süresi giriş boyutu büyüdükçe nasıl değişir?
- **Alan Karmaşıklığı (Space Complexity)**: Algoritmanın kullandığı bellek miktarı giriş boyutu arttıkça nasıl değişir?

Big-O notasyonu en çok şu şekillerde kullanılır:

- **O(1)**: Sabit zaman, yani algoritma girişin boyutundan bağımsız olarak aynı sürede çalışır.
- **O(n)**: Doğrusal zaman, yani algoritma girdinin boyutu kadar işlem yapar.
- **O(n²)**: Karesel zaman, yani algoritmanın çalışma süresi girdinin boyutunun karesiyle orantılıdır.
- **O(log n)**: Logaritmik zaman, yani algoritma girdinin boyutuna logaritmik bir oranda işlem yapar.

Bu notasyon, özellikle büyük veri kümelerinde algoritmaların verimliliğini anlamak için kullanışlıdır.

In [1]:
# Array, List, Stack, Queue, Deque

In [2]:
import array as arr

In [3]:
myArray = arr.array("i", [3,6,9,12])

In [4]:
myArray.append(15)

In [5]:
myArray

array('i', [3, 6, 9, 12, 15])

In [7]:
#LİST

myList = [1,2,3,4,5]

In [8]:
otherList = [6,7,8]

In [10]:
myList.extend(otherList)

In [11]:
myList

[1, 2, 3, 4, 5, 6, 7, 8]

Yukarıda gördüğümüz şekilde listelerde ekleme işlemi yapabiliyoruz ki hatta bir listeyi başka bir liste içerisine ekleyebiliyoruz. Bunun hafızadaki durumuna bakacak olursak 

In [12]:
result = [0] * 8

In [13]:
result

[0, 0, 0, 0, 0, 0, 0, 0]

In [14]:
type(result)

list

In [16]:
result[3] = 2
result

[0, 0, 0, 2, 0, 0, 0, 0]

Bir liste oluşturduğumuzda onun içine sonsuz şey girecek gibi bir yer ayırmıyor hafızada çünkü bir konu içerisinde bin tane liste oluşturabiliriz kod yazarken o yüzden bunun da elinden geldiğinde verimli olması lazım kod için o yüzden sys isimli kütüphaneyi import edip bu kütüphane sistem ile iletişime geçmek için kullandığımız bir kütüphane, bunu import etmek nedenimiz dizinin, listeni vs. gerçekten içeride kaç byte yer kapladığını görebilmek için 

In [17]:
import sys

In [21]:
n = 15
myDynamicArray = []

for i in range(n):
    
    myLength = len(myDynamicArray)
    myByte = sys.getsizeof(myDynamicArray)
    print(f"Length: {myLength} Byte: {myByte}")
    myDynamicArray.append(n)

Length: 0 Byte: 56
Length: 1 Byte: 88
Length: 2 Byte: 88
Length: 3 Byte: 88
Length: 4 Byte: 88
Length: 5 Byte: 120
Length: 6 Byte: 120
Length: 7 Byte: 120
Length: 8 Byte: 120
Length: 9 Byte: 184
Length: 10 Byte: 184
Length: 11 Byte: 184
Length: 12 Byte: 184
Length: 13 Byte: 184
Length: 14 Byte: 184


yukarıda gördüğümüz gibi hemen ilk başta 184 bytelık bir yer ayırmıyor hafızada yavaş yavaş ve optimizde şekilde eleman eklendikçe hafızadaki yeri de açıyor. Dikkat edersek her bir eleman için hafızayı yenilemiyor yani burada yeni bir byte kazanmak demek hafıza da aynı yerde tutmuyor bunu var olanları alıp kopyalayıp yeni bir yere taşıyor buradakileri yok ediyor sonra buradakileri garbage collector yani çöp toplayıcıları topluyor bir süre sonra, yani işte dinamik dizilerde yani listelerde bu şekilde hafızada yer alıyor

#### CONTAINS DUPLICATE sorusu

*Contains duplicate sorusu burada diyor ki :
içerisinde integerlar olan bir array vereceğiz adı nums, eğer herhangi bir değer 2 kez varsa o zaman True döndürün eğer yoksa false döndürün yani problemimiz basit* bunları çözerken time complexitiysi ne space complextyisi ne ? ne kadar ? bunlara bakmalıyız çünkü mğlakatlarda bunlardan bahsetmek zorunda kalacağız mesela ben burada O(n) "order of n" yapıyorum ya da O(1) "order of 1" yapıyorum gibi bahsetmemiz gerekecek. O yüzden kendi kendimize çözümü üretirken zaman ve yer kompleksitesini de düşünmemiz gerek benim çözümüm nasıl diye düşünmeliyiz ! yani yaptığımımz çözüme hakim olmalıyız

In [24]:
# düşünmesi en kolay olan brute force yani bunda her bir elemanı alıp diğerlerini tek tek gezerek karşılaştırmak

myList = [1,2,3,4,5,6]

In [25]:
for i in range(0,len(myList)):
    for j in range(0, len(myList)):            
        print(i,j)

0 0
0 1
0 2
0 3
0 4
0 5
1 0
1 1
1 2
1 3
1 4
1 5
2 0
2 1
2 2
2 3
2 4
2 5
3 0
3 1
3 2
3 3
3 4
3 5
4 0
4 1
4 2
4 3
4 4
4 5
5 0
5 1
5 2
5 3
5 4
5 5


Böyle yaptığımızda bu zaman olarak 0(n^2) ye denk geliyor. Ama yer olarak O(1) aynı dizi içerisindeyiz hafizada herhangi bir şey yapmadık o yüzden yer olarak çok verimli bir şey 

Time -> O(n^2)   
Space -> O(1)

Mülakatta bu çözüm kabul edilecektir ve büyük ihtimal size şunu soracaklardır peki bu daha iyi bir zaman komplesitesinde çözülebilir mi ? cevap evet 

En güzel yol burada dizi sıralanmamış olabilir ben bunu sıralarsam burada 2 değişken tutabiliriz birine x deriz birine y, bunlara pointer adını veririz genelde her seferinde x ve y'yi bir yana kaydırırız yani her seferinde bunların indexini bir arttırarak götürürüz ve her değişiklikte bunlar yan yana olacağı ve zaten sıralı olduğu için aynı ise değerler yan yana gelecektir işte bunu yaparak sadece 1 defa her yeri gezerken O(n)'de hafızada da yeni bir şey oluşturmadan yapabilir miyim ? bize problemde dizinin sıralı olduğunu söyleseydi o zaman gerçekten de bu şekilde yapabilirdik ama sıralı olmadığı için O(n)'de bunu yapamam önce sıralamam gerekiyor sonra bununla yaparım.

Sıralama algoritmalarında en ideal zaman **O(NlogN)** demiştik bu karşımıza çok çıkar o yüzden iyi öğren !!!
burada space de **O(1)** olacaktır


peki zamanı daha da düşürmek istiyorum yani **O(n)**'e düşürmemiz isteninrse ? ve yerden feraget edebilirsin yani illa sabit bişey de yapmana gerek yok bunu da **O(N)'** de yapabilirsin derlerse 
Aslında bütün bunlar arasında en kolay olan bu ben listemi hashset'e çevirisem (hashset içerisinde her değerden 1 tane tutuyor sadece) 

In [26]:
myList = [1,2,3,4,5,6,7,8,2]

In [30]:
def solution(list):
    hashSet = set()
    for num in list:
        if num in hashSet:
            return True
        hashSet.add(num)
    return False

In [31]:
solution(myList)

True

In [32]:
mylist2 = [1,2,3]

solution(mylist2)

False

In [37]:
def solution2():
    return len(myList) != len(set(myList))

# eleman sayısı farklı olacak o yüzden çok daha basit bir çözüm 

In [35]:
solution2()

True

### Single Number sorusu


" *Given a non-empty array of integers nums, every element appears twice except for one. Find that single one. You must implement a solution with a linear runtime complexity and use only constant extra space* "

Boş olmayan integer bir array veriliyor bu sefer eleman 2 kere değil de her eleman 2 kez var ama bir tanesi mutlaka single olacak yani elemanların bi tanesinden 1 tane olacak mutlaka hatta 1 elemanlı olabilir sonuçta bi tane elemandan mutlaka 1 tane koşulunu sağlıyor. Non-empty dediğine göre boş bir dizi vermez bize ama içerisinde tek eleman olabilir bu tazr durumlara biz **edge case** diyoruz yani sanki uçurumun kenarındaymış gibi kırk yılda bir karşımıza çıkabilecek bir durum :) 
Ve edge case'lerde genellikle karmaşık hesaplamalar yapmamıza gerek yok, bu tarz durumlarda örneğin bir edge case'miz var o da içerisinde tek bir eleman olması 

if len(nums) = 1:                                                                  
    return nums[0] 
    
gibisinden bir kod yazarak başlayacağız algoritmaya diyebiliriz bu da edge case'leri ele alma olarak geçer. Bu problem için çok önemli değil ama bu aklımızda olsun.
Bu problemde zor olan yer verdikleri constraitler yani verdikleri limitler sebebiyle hiç de kolay değil neden çünkü hem linear runtime da yapacaksın diyor hemde constant extra space kullanacaksın diyor yani diyor ki time complexity O(N) olacak space complexity'nde O(1) olacak diyor. 

İlk akla gelen ile brute force yapabiliriz yani her bir elemanı alıp diğerleri ile karşılaştırabiliriz space **O(1)** ama bunun zamanı **O(N^2)** o yüzden olmaz.


Peki sıralayıp yapabilir miyim ? yine pointer koyup tek tek bunları gezerek bulur gerçekten ama dizerken yine **O(NlogN)** olacak o yüzden olmaz yine O(nlogn) olduğu için O(N)'den daha yukarıya bir zaman oluyor lineer zaman olmuyor linear ve logaritmiğin karışımı oluyor o yüzden yine bizim işimize yaramayacak 


Burada da hashset ya da hashmap kullanarak bu sorunu çözebiliriz fakat yine tek tek buradakileri bri loopa sokarız alırız diğer bir kolaksiyona koyarken kontrol ederiz daha önce varsa onu çıkartırız hem yenisini koymayız hem eskisini çıkartırız böylece O(N) zamanda bitirmeyi başarabiliriz fakat bu sefer zamandan tasarruf ederken spacei de yine O(N)'e götürmüş oluruz.

Peki hem ekstra bir yer kullanmadan hemde zamanı O(N)'de tutarak nasıl bunu yapabilirim ?

Burada **BİT MANİPÜLASYONUNU** kullanıyoruz. Öncelikle 0-1 mantığını hatırlayalım örneğin;           
504 ün ondalık gösterimi -> 5x10^2, 0x10^1, 4x10^0

aynı şey binary sayı sistemi içinde geçerli !!              
110101 binary gösterimi -> 1x2^5 + 1x2^4 + 0x2^3 + 1x2^2 + 0x2^1 + 1x2^0  = 32 + 16 + 0 + 4 + 0 + 1 = 53

110101 = 53 


Bizim sorumuzda karşımıza çıkan rakamlar decimal olarak ifade edilebilir. Mantık kapılarından hatırlayacağımız üzere XOR mantık kapısının bir özelliği vardı "eğer kendisiyle yani 0-0 ya da 1-1 'i XOR kapısına sokarsanız 0 elde edersiniz sadece ve sadece 2 girdi birbirinden farklı ise  1 elde edersiniz " ve diğer dikkat edilmesi gereken konu 0 hiçbir şeyi değiştirmiyor 0 ne ile XOR kapısına girerse o oluyor. İşte pythonda bunu yapabildiğimizde kolaylıkla bu sorunu çözebiliriz.

Örneğin [4,1,3,1,3] şeklinde bir dizi olsun Ben bir tane değişken alırım X değişkeni olsun x'i 4 olarak atarım x=4 veya ilk başta 0 olarak da atayabilirim ve sonrasında tek tek bunlarda XOR a sokarım, 0 olarak atadığım için örneğin 3 ile XOR a soktuğumda sonuç 3 çıkacaktır sonra 3 ile 3 ü XOR yaptığımda sonuç 0 çıkacaktır, 0 ile 1 i XOR yaptığımda sonuç 1 çıkacaktır, 1 ile 1 i XOR yaptığımda sonuc 0 çıkacaktır eninde sonunda her rakam bu değişken ile XOR a tabi tutulursa bunların aynı olanları birbirini eleyecektir geriye 1 tane rakam kalacaktır farklı olan o rakamda bu örnekte 4 olacaktır. 
Yani sadece 1 defa geçtiğim için listeden O(N) de yapıyorum ve yeni bri hafıza da harcamadığım için O(1) de olur.

**işte buna BİT MANİPULATİON denir ve önemlidir !!!**

 ptyhonda XOR işlemi **^** işareti ile yapılır

In [40]:
def singleNumber():
    myList = [3,5,2,2,4,3,4]
    
    result = 0
    
    for num in myList:
        result = num ^ result
    return result

In [41]:
singleNumber()

5

Burada şu şekilde düşünebilirsiniz "ya 3 ile 5 XOR a giriyor nasıl ben anlamadım " burada 3 ile 5 XOR a girmiyor 3 -> 11 
 5 -> 101 olarak binary karşılıkları XOR a giriyor, ve eninde sonunda aynı birler ve aynı sıfırlar birbirini götürüyor sadece tek 1'ler kalıyor ve o birler bu örnekte 5 için nasıl diziliyorsa o bizim sonucumuz oluyor 
 
 XOR işlemi (^) bit düzeyinde yapılan bir işlemdir. İki sayıyı binary (ikili) sayı sisteminde karşılaştırarak çalışır ve her bir bit çiftini şu kurallara göre işler:

- Eğer iki bit aynıysa, sonuç **0** olur.
- Eğer iki bit farklıysa, sonuç **1** olur.

Yani, `a ^ b` işlemi iki sayının binary (ikili) hallerini karşılaştırır ve sonucunu oluşturur.

Şimdi adım adım **5 ^ 3 = 6** işlemini inceleyelim:

### 1. Sayıları Binary'ye Çevirme:
- 5 sayısının binary (ikili) gösterimi: **0101**
- 3 sayısının binary (ikili) gösterimi: **0011**

### 2. XOR İşlemi:
İki sayıyı bit bit karşılaştıralım:

```
   0101   (5'in binary gösterimi)
^  0011   (3'ün binary gösterimi)
---------
   0110   (sonuç: 6)
```

### XOR Adım Adım İnceleme:
- İlk bit: 0 ^ 0 = 0 (Her iki bit de aynı, sonuç 0)
- İkinci bit: 1 ^ 0 = 1 (Farklı bitler, sonuç 1)
- Üçüncü bit: 0 ^ 1 = 1 (Farklı bitler, sonuç 1)
- Dördüncü bit: 1 ^ 1 = 0 (Her iki bit de aynı, sonuç 0)

Sonuç: **0110** yani bu da onluk sistemde **6**'ya karşılık gelir.

### 3. Sonuç:
Bu yüzden `5 ^ 3 = 6`. XOR işlemi iki sayının ikili bitlerini karşılaştırarak sonucunu üretir.

Umarım bu açıklama XOR işlemini daha iyi anlamanıza yardımcı olmuştur!

### MAJORITY ELEMENT sorusu

*"Given an array nums of size n ,return the majority element. The majority element is the element that appears more than [n/2] times. you may assume that the mojarity element always exists in the array"*

Burada bir dizi içerisindeki baskın eleman ne onu bulmak bize bir dizi veriliyor ve bu dizi içerisinde baskın eleman ne onu bul diyo majority elemanı da şu şekilde tanımlıyor örneğin içerisinde 10 tane eleman varsa 5'i aynı olan eleman n/2 si yani. Ve şunu diyo her zaman %50 sinden fazlası o eleman olacak olmadığı durumları if le kontrol etmene gerek yok diyo


constraints:
* n == nums.length
* 1 <= n <= 5 * 10^4
* -10^9 <= nums[i] <= 10^9

constrain yani kısıtlamalar kısmında verir neler olduğu burada n nin yani eleman sayısının sonsuz olmadığını belirli bir aralıkta olduğunun ve nums elemanlarının da  belirli bir aralıkta değer alabileceğini belirtiyor. Yani integer değerinin alabilir 32 bitten daha yukarısını düşünmeye gerek yok 

ve mülakatçı şunu sorabilir linear time O(N) ve space O(1) yapabilir misin diye demek ki O(N) ve O(1) yapabilecek bir çözüm var !!

[1,1,1,1,2,3,4] dizimiz bu şekilde ve bizden majority eleman olan 1'i bulmamız isteniyor. Bunları bir sözlüğe ya da hashmape atarım.
Bunu sözlüğe atarsam anahtarlar integer olur, değerler de kaç tane olduğu olur. {1:4,2:1,3:1} ve her zaman buradaki rakamları takip ederim bir değişken atarım maxnumber diye bu her bir döngüde yani her bir loop da bundan daha büyük bir rakam gelmiş mi onu kontrol ederim ve en sonunda bundan daha büyük bir rakam gelmediyse bu hangi anahtara aitse onu alırım demek ki majorty buymuş diye

In [47]:
myList = [2,3,4,1,1,1,1,1,5,3,2]

In [45]:
def findMajority():
    numbers = {}
    result = 0
    maxNumber = 0
    
    for num in myList:
        # numbers[num] = 1 + numbers[num]  eğer sözlükte daha önce böyle bir anahtar eklenmediyse bu hata veriyor 
        numbers[num] = 1 + numbers.get(num,0) # bu değerde bir anahtar varsa ona ekle yoksa ekle ve default değerini 1 yap
        
        if numbers[num] > maxNumber:
            result = num
        maxNumber = max(maxNumber, numbers[num])
        
    return result
        

In [48]:
findMajority()

1

Böyle problemleri **Bayer Moore algoritması** ile çözüyorlar bu algoritma önemli ve güzel bir algoritma göz atmakta fayda var.

biraz daha zor bir dizi üzerinden gidelim :

[3,4,4,3,3,3,4] bir result değişkeni ve count ile işlemlerimize başlayalım ;

* ilk eleman olan 3 için -> result = 3, count = 1 (result dan kaç tane olduğunu sayıyoruz)

* 4 için -> result = 3, count = 0  (eleman değiştiğinde o anki result değerinden farklı bir eleman geldiğinde count değerini 1 eksiltiyoruz eğer count = 0 ise result değerini o anki gelen değer yapıyoruz ve o değeri saymaya başlıyoruz)

* 4 için -> result = 4, count = 1

* 3 için -> result = 4, count = 0

* 3 için -> result = 3, count = 0 olduğu için resultı değiştir ve count = 1 yap

* 3 için -> result = 3, count = 2

* 4 için -> result = 3, count = 1

SONUÇ olarak majority değer 3 olarak karşımıza çıkıyor 

!! **DİKKAT bu tarz problemlerin bu çözümünde  en az %50 majority şartı olmalı**


In [49]:
def boyerMoore():
    result = 0
    count = 0
    
    for num in myList:
        if count == 0:
            result = num
        if num == result:
            count += 1
        else:
            count -= 1
    return result

In [50]:
def boyerMoore():
    result = 0
    count = 0
    
    for num in myList:
        if count == 0:
            result = num
        count += 1 if num == result else -1
        
    return result

In [51]:
boyerMoore()

1

Burada bu algoritma ile yine ben majoriteyi buldum ama yeni bir hashmap yeni bir set vs oluşturmadığım için ekstradan hafıza kullanmadım yani özellikle majorite problemlerinde bu algoritma aklımızda olsun bu soru facebook da çok soruluyormuş